src/installer-generator/installer-generator.c

```diff
@@ -0,0 +1,156 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "blockdev-util.h"
+#include "device-util.h"
+#include "dropin.h"
+#include "generator.h"
+#include "initrd-util.h"
+#include "parse-util.h"
+#include "path-util.h"
+#include "proc-cmdline.h"
+#include "special.h"
+#include "volatile-util.h"
+
+#define REPART_SERVICE "systemd-repart.service"
+
+static const char *arg_dest = NULL;
+static enum {
+        INSTALLER_NO,
+        INSTALLER_YES,
+        INSTALLER_AUTO,
+} arg_enabled = INSTALLER_NO;
+static bool arg_disable_repart = true;
+static VolatileMode arg_volatile = VOLATILE_YES;
+
+static int parse_cmdline(const char *key, const char *value, void *data) {
+        int r;
+
+        assert(key);
+
+        if (proc_cmdline_key_streq(key, "systemd.installer")) {
+                if (!value)
+                        arg_enabled = INSTALLER_YES;
+                else if (streq(value, "auto"))
+                        arg_enabled = INSTALLER_AUTO;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer value \"%s\", ignoring: %m", value);
+                        arg_enabled = r ? INSTALLER_YES : INSTALLER_NO;
+                }
+        } else if (proc_cmdline_key_streq(key, "systemd.installer_disable_repart")) {
+                r = value ? parse_boolean(value) : true;
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse systemd.installer_disable_repart value \"%s\", ignoring: %m", value);
+                else
+                        arg_disable_repart = r;
+        } else if (proc_cmdline_key_streq(key, "systemd.installer_volatile")) {
+                if (!value)
+                        arg_volatile = VOLATILE_YES;
+                else if (streq(value, "overlay"))
+                        arg_volatile = VOLATILE_OVERLAY;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer_volatile value \"%s\", ignoring: %m", value);
+                        else
+                                arg_volatile = r ? VOLATILE_YES : VOLATILE_NO;
+                }
+        }
+
+        return 0;
+}
+
+static int conditionalize_unit(const char *unit) {
+        return write_drop_in(arg_dest, unit, 50, "installer-auto",
+                             "# Automatically generated by systemd-installer-generator\n\n"
+                             "[Service]\n"
+                             "Environment=SYSTEMD_INSTALLER_DISABLE_IF_REMOVABLE=1\n");
```

Comment:
hmm, my thinking was that if we are in installer mode, then we'd write out a drop-in for repart that does something like this:

```
[Service]
ExecStart=
ExecStart={{ROOTBINDIR}}/systemd-repart --dry-run=no -C installer
```

which is a short way to say:

```
[Service]
ExecStart=
ExecStart={{ROOTBINDIR}}/systemd-repart --dry-run=no --definitions=/usr/lib/sysupdate.installer.d/ --definitions=/etc/sysupdate.installer.d/ --definitions=/run/sysupdate.d/
```

The net effect is that repart will read its definitions from a different dir. And if that dir doesn't exist it just becomes a NOP and doesn't do anything.

I think that's a lot more powerful, since it means that you can still run repart stuff if you want on a live medium, for example because even on a live medium it might make sense to adjust to the medium.